spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/r2dbc/ConnectionFactoryConfigurations.java

```diff
@@ -86,13 +97,43 @@ ConnectionPool connectionFactory(R2dbcProperties properties, ResourceLoader reso
 
 		@Bean
 		ConnectionFactory connectionFactory(R2dbcProperties properties, ResourceLoader resourceLoader,
-				ObjectProvider<ConnectionFactoryOptionsBuilderCustomizer> customizers) {
+				ObjectProvider<ConnectionFactoryOptionsBuilderCustomizer> customizers,
+				ObjectProvider<ProxyListenersProvider> proxyListenersProviders) {
 			return createConnectionFactory(properties, resourceLoader.getClassLoader(),
-					customizers.orderedStream().collect(Collectors.toList()));
+					customizers.orderedStream().collect(Collectors.toList()),
+					proxyListenersProviders.getIfAvailable(() -> ProxyListenersProvider.EMPTY));
 		}
 
 	}
 
+	@Configuration(proxyBeanMethods = false)
+	@ConditionalOnClass(ProxyConnectionFactory.class)
+	@ConditionalOnProperty(prefix = "spring.r2dbc.proxy", value = "enabled", havingValue = "true",
+			matchIfMissing = true)
+	static class Proxy {
```

Comment:
It would be nice if it didn't have to surface that way. this should register a  customizer implementation that does the necessary thing. And if no listeners are present then the customizer does nothing.

Reply:
Currently, `ConnectionFactoryOptionsBuilderCustomizer` provides a way to customize `Option`s on `ConnectionFactoryBuilder`. So, I could use it as a hook point to add/replace some of the `Option`s.

I wasn't sure it is ok to use customizer facility within Spring Boot, but if that's ok, then here can define a customizer bean (probably with `@Order` to allow users to run own customizers before and after). Then, it will be pretty much all we need for `r2dbc-proxy`.

Reply:
Yeah I think that's ok.

Reply:
I still use this `Proxy` configuration class but now it has only single processor(`Function`) bean that is used by `ConnectionFactoryConfigurations#createConnectionFactory`.

In the latest impl, the conditions to apply proxy are:
- `spring.r2dbc.proxy.enabled` is `true`
- `spring.r2dbc.url` does not contain `:proxy:`
- `ProxyExecutionListener` beans OR a `ProxyConfig` bean exist


I have explored the option to use `ConnectionFactoryOptionsBuilderCustomizer` but the connection factory `Option` is limited to the use of connection factory discovery mechanism.

Currently, similar to auto configuration for `r2dbc-pool`, auto config for `r2dbc-proxy` does not modify connection-url. When `spring.r2dbc.url` contains `proxy` prefix, it will backoff the auto configuration. (same as pool)
So options will not be picked up even if they are customized by auto configuration.